main/src/ca/uwaterloo/flix/language/phase/EntryPoint.scala

```diff
@@ -18,264 +18,564 @@ package ca.uwaterloo.flix.language.phase
-      case None => root.defs.get(DefaultEntryPoint) match {
-        case None => None
-        case Some(entryPoint) => Some(entryPoint)
+      case None =>
+        root.defs.get(defaultMainName) match {
+        case None =>
+          // No main is given and default does not exist - no main.
+          (root, Nil)
+        case Some(entryPoint) =>
+          // No main is given but default exists - use default.
+          (root.copy(entryPoint = Some(entryPoint.sym)), Nil)
       }
       case Some(sym) => root.defs.get(sym) match {
+        case Some(_) =>
+          // A main is given and it exists - use it.
+          (root, Nil)
         case None =>
-          val error = EntryPointError.EntryPointNotFound(sym, getArbitrarySourceLocation(root))
-          sctx.errors.add(error)
-          None
-        case Some(entryPoint) => Some(entryPoint)
+          // A main is given and it does not exist - no main and give an error.
+          (root.copy(entryPoint = None), List(EntryPointError.MainEntryPointNotFound(sym)))
       }
     }
   }
 
   /**
-    * Retrieves an arbitrary source location from the root.
+    * A function is an entry point if:
+    *   - It is the main function (called `main` by default, but can configured
+    *     to an arbitrary name).
+    *   - It is a test (annotated with `@Test`).
+    *   - It is an exported function (annotated with `@Export`).
     */
-  private def getArbitrarySourceLocation(root: TypedAst.Root): SourceLocation = {
-    root.sources.headOption match {
-      // Case 1: Some arbitrary source. Use its location.
-      case Some((_, loc)) => loc
-      // Case 2: No inputs. Give up and use unknown.
-      case None => SourceLocation.Unknown
-    }
-  }
+  private def isEntryPoint(defn: TypedAst.Def)(implicit root: TypedAst.Root): Boolean =
+    isMain(defn) || isTest(defn) || isExport(defn)
+
+  /** Returns `true` if `defn` is a test. */
+  private def isTest(defn: TypedAst.Def): Boolean =
+    defn.spec.ann.isTest
+
+  /** Returns `true` if `defn` is an exported function. */
+  private def isExport(defn: TypedAst.Def): Boolean =
+    defn.spec.ann.isExport
 
+  /** Returns `true` if `defn` is the main function. */
+  private def isMain(defn: TypedAst.Def)(implicit root: TypedAst.Root): Boolean =
+    root.entryPoint.contains(defn.sym)
 
   /**
-    * Checks that the given def is a valid entry point,
-    * and returns a new entry point that calls it.
+    * Returns `true` if `tpe` is equivalent to Unit (via type aliases).
     *
-    * The new entry point should be added to the AST.
+    * N.B.: `tpe` must not have type variables or associated types.
     */
-  private def visitEntryPoint(defn: TypedAst.Def, root: TypedAst.Root, traitEnv: TraitEnv)(implicit sctx: SharedContext, flix: Flix): TypedAst.Def = {
-    checkEntryPointArgs(defn, traitEnv)
-    checkEntryPointResult(defn, root, traitEnv)
-    mkEntryPoint(defn, root)
+  @tailrec
+  private def isUnitType(tpe: Type): Boolean = tpe match {
```

Comment:
Do we want this? Seems like wacky town to allow
```
type alias JonathanLindegaardStarup = Unit
def main(_: JonathanLindegaardStarup): Int32
```

Reply:
hmm yeah probably nobody will notice either way. I do like keeping it correct though since you can stepwise go through

```
def main(): Int32 = ...
def main(_: Unit): Int32 = ...
def main(_: UnitAlias): Int32 = ...
```
and not supporting the last one makes the check feel syntactic and shallow

Reply:
I am OK either way.